docker_node: Docker Swarm node operations module

WojciechowskiPiotr opened the pull request:
##### SUMMARY
New module to handle Docker Swarm node operations. It must be executed on Swarm `manager` node and allows changing the node availability, role, and labels (fixes #49547). Part of the commit is also a shared library for all common methods for other Swarm modules.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
docker_node

##### ADDITIONAL INFORMATION
This PR is a copy of #50428, required due to the wrong branch in original PR. The open discussion in old PR is on module name and where put some inspect features, either in `docker_node` or `docker_node_facts` (#50571). The upcoming `docker_swarm_facts` will use the shared code from this PR. 
This module is equivalent of `docker node` CLI command allowing now changing the node availability and role as well as applying or removing the labels from the node. It must be executed on Swarm manager node otherwise exception is raised. 
When RP is merged into `ansible/devel` then we can update `docker_swarm` module so it uses the shared methods from `module_utils/docker_swarm` instead of own local wherever it is possible.

Commit:
* docker_node: New module for operations on Docker Swarm node
* Shared code for Docker Swarm modules

WojciechowskiPiotr commented:
ansible/community#408

Commit:
* docker_node: Removed the attribute `force` as it is not used for any operation

WojciechowskiPiotr commented:
ready_for_review

Would be nice to get this PR reviewed and hopefully merged in parallel to #50571 as those are not in conflict to each other

felixfontein commented:
I think it makes more sense to first get #50571 merged, and then modify `docker_node_facts` to use the new module_utils as well (in this PR) -- as well as discuss whether the `list` mode should be moved to `docker_node_facts` (I'm kind of in favor of this).

Anyway, to get something merged, some more reviews are needed. Anyone? :)

Commit:
docker_node_facts: Reverting to repository version, moving this change to separate PR

felixfontein commented:
Now that #50571 has been merged, can you rebase this PR against `devel` and make `docker_node_facts` use the new module_utils as well?

Commit:
docker_node_facts: Update module to use client class AnsibleDockerSwarmClient instead of AnsibleDockerClient

WojciechowskiPiotr commented:
The change in `docker_node_facts` is purely cosmetic, already committed in my repo, will be part of next push to this PR.

As @tbouvet agreed with me in #50428, we will use the `docker_node` as the module name. That would put changing the name of `docker_service` into `docker_compose` in more flavor so we can later rename `docker_swarm_service` into `docker_service`. It will take six releases, but I think it is good for modules convention sanity in `cloud/docker`. 

As for the `list` as I said before in #50428 only the `docker_image_facts` does not require `name` parameter, and it accepts a list of strings. Other modules require a single string as `name`. The CLI commands aways allows a list of objects to inspect (empty list not allowed) and provides the `ls` action to just list objects with basic information. I like the approach where the `_facts` module is like an `inspect` and `non-_facts` have an implementation of `ls` command.

I think the options to consider are:

1. Leave `docker_node_facts` as it is, so it returns information on a single node, adjusts the `docker_image_facts` to this approach, create the `ls` command implementation in an operational module, it may cover the filters as well
- Pros: Easy `docker_*_facts` modules in code, the single way you can use them in playbooks
- Cons: Playbook developer need to know the name of the object to inspect every time, so there should be some kind of implementation of CLI command `docker [network,image,container,etc] ls` in each module to give a list of objects
- Consistency with CLI - partial

2. Let parameter `name` optional in all `docker_*_facts` modules and accept the list as a parameter, if `name` not provided then return information about all objects.
- Pros: Easy way to get information on every object, specific objects or a single object, shouldn't be hard to implement. May include filters too.
- Cons: If called without a `name` parameter it returns lots of information, and later the playbook code may not be so clear due to big multi-levels structures user have to parse. If it covers the filters as well, it will increase the complexity of the facts modules
- Consistency with CLI - partial

3. Let parameter `name` optional in all `docker_*_facts` modules and accept the list as a parameter, if `name` not provided then return information about all objects but also add the `ls` command output to the operation modules.
- Pros: Easy way to get information on every object, specific objects or a single object, shouldn't be hard to implement. The `_facts` will not allow additional filtering, this feature will be implemented as part of the `non-_facts` module.
- Cons: If called without a `name` parameter it returns lots of information and later the playbook code may not be so clear due to big multi-levels structures user have to parse. If it covers the filters as well, it will increase the complexity of the facts modules
- Consistency with CLI - full

4. Let parameter `name` required as list of `strings` and implement the `ls` method in each non-facts module. This way `facts` modules will return the full information on at least single object, while `ls` command implementation returns limited information and can cover filters
- Pros: Playbook developer will call the `facts` module to inspect known objects, because it will be a list it will be possible to get information in one task, the `ls` command implementation in each module may return the structure identical to the CLI, so limited in number of information, it can also cover the filters
- Cons: Need a `ls` command implementation in each module
- Consistency with CLI - full

One more thing worth noticing is the difference between Swarm and non-Swarm hosts. The non-Swarm modules (`image`, `container`, `volume`, `network`) you need to run on every host in inventory while swarm modules (with some exceptions of node operations not proposed yet in this PR) have to be run on hosts acting as Swarm Managers. This role is dynamic by its nature and promoting/demoting nodes shouldn't be unusual operation to keep a quorum in the cluster. Also, each manager will return same information about the cluster and its components (node, config, secret, other swarm modules) so it will be a waste of time and resources to run facts module on each node from inventory every time.

With the approach in point 4) the playbook developer may fetch the nodes lists and roles from `docker_node` using `state: list` to get a simple structure with just 6 or 7 attributes as in `docker node ls` command output, he may use additional filters as well. It is easy to use the list if playbook runs on localhost and uses the API. With SSH and hosts in inventory dynamic states in Swarm are not so easy to handle when you try to minimize number of tasks in playbook, time to run it and resources it will require. 

To sum this up - I like the consistency between the modules and I like the idea that modules are working in similar way as CLI because most people will first know the CLI then will try automate the tasks. My preference is option **4** in this case, but I am happy to go with approach most of the docker modules maintainers here will recommend as preferred.

felixfontein commented:
There are a few things to consider:
1. Normal modules should do something (except for check mode), and `_facts` modules should query stuff without modifying. Having data querying functionality in normal modules is discouraged. I'm definitely against adding a `list`/`ls` method to any non-`_facts` module. If there is `list`/`ls` support, it should be handled by the `_facts` module (or by a new module).
2. Ansible should not mirror exactly what can be done via CLI or API, but what is useful. Is there a concrete need for getting a list of containers, images, ...? I can see that it makes sense for Swarm nodes, but I doubt it for other objects.

Also, for offering lists in a `_facts` module, I would prefer having a choices option `results` with default value `single`, and alternative choice `list` (names open for bikeshedding as usual ;) ). For `single`, `name` would be required; for `list`, it would not be (and will be ignored).

WojciechowskiPiotr commented:
1. This is a good argument against my approach and I have no problem in going this way
2. It is not a necessity to mirror but if it's not against the general usage then why it shouldn't? 

You mean a list in input or output? As input I think best would be as it is in `docker_image_facts` plus if `name` is missing then it returns facts about all objects.

felixfontein commented:
About 2: because it increases complexity, both for the module's code and for the module's documentation. If there's a good enough usecase, I don't mind having it, but having it just because we can doesn't seem like a good idea for me.

I meant lists as output. I don't think there's a use for having lists as input; one can use task-level loops if one really has one. (At least, I can't think of a good use for having lists as input. And I wouldn't mind getting rid of the lists input for `docker_image_facts`; unfortunately it has been there already for a longer time...)

felixfontein commented:
Do you mean `docker_container_facts` or `docker_node_facts`? And why not use Ansible's task-level loop facility for such a use-case?

WojciechowskiPiotr commented:
In `docker_node_facts`. Both input list and filters are meant to limit the number of information you want later to process in the playbook. Of course, there is nothing wrong in getting information on all registered nodes, then in the next tasks loop through the results, make conditionals etc but that increases the complexity of playbook and usually makes it harder to read. 

If you look at other groups in the `cloud` most `_facts` modules allows limiting the scope by `filters`, or `labels`, sometimes more vendor-specific attribute. Especially the AWS, Azure, GCP or OpenStack modules

Commit:
docker_node_facts: List of nodes can be provided as input, inspect all registered nodes or manager host itself

felixfontein commented:
I didn't say that `docker_node_facts` cannot have a list output; for me, that's the only `docker_*_facts` module where I see that it is useful to have that functionality.

For me, the `*_facts` modules query information about a specific object: `docker_container_facts` about a docker container, `docker_volume_facts` about a docker volume, etc. So for me, `docker_node_facts` gets facts about a specific node. That's why I would suggest to add two modes to it: one for querying a specific node (with `exists` information being returned, and where an input list makes no sense), and one for listing nodes (where you can filter and whatever to get a list of nodes in the end). Since the functionality is so disjoint, how about having it in two different modules? The list functionality could also be a part of `docker_swarm_facts`, since listing nodes is something you can get from a swarm (and not from a node).

(I checked some random openstack modules and didn't saw any list input. Also, the GCP modules are special since they are auto-generated from the API. In the AWS modules I saw indeed some with list input. It really depends on what you want to do with the queried information. Whether you are interested in one specific object, or whether you want to list and filter.)

WojciechowskiPiotr commented:
@felixfontein @tbouvet - Moving the `list` from `docker_node` somewhere else is not a problem. 

I've pushed an update to `docker_node_facts`. The `name` is now optional, if missing then facts about all registered nodes are available in output. It also accepts in `name` a list to query about single or multiple nodes. Additionally, the `self` to get node facts about the host on which module is run on (useful if you use SSH as connection method so you can in one task get facts about all swarm managers from inventory). Requires some polishing before it is final

The whole `list` features I will implement in `docker_swarm_facts`, it will include filters

WojciechowskiPiotr commented:
Some thoughts about `docker_node` now.

`state: update` - update the node parameters (role, availability, labels defined as additional parameters), must be run on swarm manager. The parameters that are not specified in task remains unchanged. The promotion and demotion between `manager` and `worker` roles on CLI can be done via CLI commands `docker node promote` and `docker node demote` but I am happy to leave all in `state: update` for clarity.
And at this moment it will be the only state, might be defined as default giving the possibility to have more states in future depending on new features in docker itself. It also leaves joining and leaving as well as removing `down` nodes as operations in `docker_swarm` module.

The `docker_swarm_facts` have to include swarm inspection (the `state: inspect` in current `docker_swarm`), swarm node lists (`docker node ls`) with filters and swarm node tasks lists `docker node ps`) with filters.

Any thoughts?

Commit:
docker_node: Update in method name called from AnsibleDockerSwarmClient

Commit:
docker_node: Additional method to shared module to get formatted output list of registered nodes

Commit:
docker_node: Additional method to shared module to get formatted output list of registered nodes

Commit:
docker_node: removed state list (featue moved to docker_swarm_facts)

Commit:
docker_node: Node labels manipulation (remove, replace, merge)

Commit:
module_utils/docker_swarm: Updated output for nodes list - adding swarm leader flag

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node_facts.py:
```diff
@@ -70,12 +96,14 @@
+      - Will contain multiple entries if more than one node provided in I(name).
+      - If I(name) contains list of nodes, the output will contain information only about registered nodes.
     returned: always
     type: dict
```
That's no longer true. You return a `list`. Also, the name is misleading, since these are not facts about one node, but a list of facts of nodes. Maybe use `nodes_facts`, or simply `nodes`?

Also, `exists` doesn't make sense any more when this is a list.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node_facts.py:
```diff
@@ -29,8 +29,19 @@
+    required: false
+  self:
+    description:
+      - If C(True) then query the host module run on or one specified as I(docker_host).
```
```suggestion
      - If C(true), queries the node (i.e. the docker daemon) the module communicates with.
```

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node_facts.py:
```diff
@@ -29,8 +29,19 @@
+  self:
+    description:
+      - If C(True) then query the host module run on or one specified as I(docker_host).
+      - If C(True) then I(name) is ignored.
```
In Ansible, `true` or `yes` is preferred over other boolean forms. (Same for `false`/`no`.)

([Reference](https://github.com/ansible/community/issues/389#issuecomment-452327393))

WojciechowskiPiotr replied:
It says both yes/no and true/false are accepted with no preference to any in documentation and examples. I don't mind changing it of course to yes/no here.

I also noticed in `docker_swarm` module there is inconsistency in using boolean values in documentation vs. examples

felixfontein replied:
You're currently using `True`, which is neither `true` nor `yes`. You can use either `true` or `yes`, but not `True`.

WojciechowskiPiotr replied:
The capital letters in values as per your other comments were already amended, I got your comment wrong

felixfontein replied:
Ok, no problem :)

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,290 @@
+version_added: "2.8"
+description:
+    - Manages the Docker nodes via Swarm Manager
+    - Change node role
```
Can you combine this and the next two paragraphs into one sentence? For example, "This module allows to change the node's role, its availability, and to modify, add or remove node labels."

felixfontein replied:
(Also, please use a period at the end of a sentence.)

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,290 @@
+    - Change node availability
+    - Change or remove node labels
+options:
+    state:
```
If it only has one value (and there's no obvious extension coming up), please remove it.

WojciechowskiPiotr replied:
I was thinking about this and left it for discussion. You are right `update` is the only action now, and no extensions are on the roadmap at the moment, but if there will be any in future, we don't have to add the `state` later. Also, if we remove the `state` then more accurate name for the module might be `docker_node_update`, but that will go against the naming convention.

An alternative approach might be using the `state` instead of the `availability` - all module parameters would be optional and those not specified won't be changed in the upate.

felixfontein replied:
Re-adding state is trivial (by using the correct `default` value).

felixfontein replied:
> An alternative approach might be using the `state` instead of the `availability` - all module parameters would be optional and those not specified won't be changed in the upate.

I don't understand what you want to say with this. This here: "all module parameters would be optional and those not specified won't be changed in the upate" is what I would expect from docker_* modules.

WojciechowskiPiotr replied:
I just meant changing `availability` to `state` but that may be a little confusing because in `docker_swarm` the `state` values of `join`, `absent` and `remove` are about node membership in the cluster, and here the 'state' would mean just availability.

felixfontein replied:
Then it's probably better to keep `availability` as it is.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,290 @@
+        description:
+            - Defines the operation on the lables assigned to node.
+            - Set to C(merge) to merge provided labels with already assigned ones. It will update existing keys.
+              The I(labels) must be specified.
```
It is unclear what this means. Do you mean that the labels specified will be added/updated, and other labels already there (but not specified here) will be left alone? (I.e. the standard behavior for other `docker_*` modules?)

WojciechowskiPiotr replied:
I updated the description with more extended information on `merge`. Also will add one more action here to allow removing specific labels defined in `label`

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,290 @@
+            - Set to C(merge) to merge provided labels with already assigned ones. It will update existing keys.
+              The I(labels) must be specified.
+            - Set to C(replace) to replace assigned labels with provided ones. The I(labels) must be specified.
+            - Set to C(remove) to remove all labels assigned to node.
```
This is redundant, one can use `replace` with `labels: {}`.

WojciechowskiPiotr replied:
Will use `remove` to remove single or multiple labels provided in `labels`

WojciechowskiPiotr replied:
Actually I think it may be better to use `remove` for both. If `labels` is not specified then it will remove all lables, otherwise it will remove just labels specified in `labels`

felixfontein replied:
I don't think that's a good UI.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,290 @@
+            - Swarm name of the node.
+            - Used with I(state=update).
+    labels:
+        description: User-defined key/value metadata.
```
Always add a type, in particular if it is not `str`! (Otherwise, it will look ugly in the documenation.)

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,290 @@
+          - remove
+        default: 'merge'
+    availability:
+        description: Node availability status to assign
```
You might want to add a sentence that the configured value will not be updated if `availability` is not specified.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,290 @@
+          - pause
+          - drain
+    role:
+        description: Node role to assign
```
You might want to add a sentence that the configured value will not be updated if `availability` is not specified.

WojciechowskiPiotr replied:
The `availability` is always specified, if not directly in playbook then the current node availability is used for API call.

felixfontein replied:
Why is that? Doesn't that exactly mean that the value is not updated when it is not specified?

WojciechowskiPiotr replied:
The `availability` and `role` always have to be included in API call, if either is missing then API call returns `400`. The Docker documentation and python `docker` module documentation and comments in the code are a little misleading so I tested it by myself

If you just call `/update` with the following body

```
{
  "Availability": "drain"
}
```
the response is APIError 400 together with error description
```
{
    "message": "invalid Role: \"\""
}
```

If you specify just the availability
```
{
  "Role": "worker"
}
```
the APIError is 400 and message
```
{
    "message": "invalid Availability: \"\""
}
```

The body of `/update` POST must at least include `Role` and `Availability` (and `Version` as a key in a call to be precise). This is checked on Docker side even if you are not updating those attributes. Only the `Label` is cleared to empty object `{}` and 'Name` removed if missing from POST request body.

In such case, if any of those two parameters are missing in Ansible module use in a playbook then values assigned to the node will be provided unchanged for the API call, this way you don't have to provide both of them in the playbook every time.

felixfontein replied:
So if you use the current values (returned by inspect), these values are not changed. That's what I meant in the original sentence that should be added (and that's usually the only thing the user cares about).

WojciechowskiPiotr replied:
Ah ok, I connected it with the identical comment to line 56 above and my understanding was that you want `availability` to be required parameter somehow

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,290 @@
+  description: Provides the actions done on the swarm.
+  returned: when action failed.
+  type: list
+  example: "['This cluster is already a swarm cluster']"
```
Is someone actually using this (except for debugging)? It might be more useful to add proper `diff` support.

WojciechowskiPiotr replied:
I left it to preserve the compatibility with outputs in `docker_swarm` module. @tbouvet ?

WojciechowskiPiotr replied:
@tbouvet - any thoughts on that? Would also appreciate a full review as for this PR there are not many doubts to close I think.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,290 @@
+            "update": self.node_update,
+        }
+
+        choice_map.get(self.parameters.state)()
```
Why not `choice_map[self.parameters.state]()`? That's shorter, and if it doesn't find the key, it will crash anyway.

WojciechowskiPiotr replied:
Except for being shorter is there any other advantage? In `docker_swarm` we have `.get()` as well, in other modules we have simple conditionals.

felixfontein replied:
Why use a longer / more complicated form when the shorter also works, and improves readability? :)

felixfontein replied:
> Why use a longer / more complicated form when the shorter also works, and improves readability? :)

WojciechowskiPiotr replied:
Tbh, in this case, in my opinion, it is not an improvement just preferred way of coding. but I don't mind changing it :)

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,290 @@
+        choice_map.get(self.parameters.state)()
+
+    def node_update(self):
+        if not (self.client.check_if_swarm_manager()):
```
Why not move this (and the next) check to the constructor?

WojciechowskiPiotr replied:
That can be done, but needs to include an exception for `state: join` in `docker_swarm` module. I understand `docker_swarm` will use the `AnsibleDockerSwarmClient` class at some point, so we can add the required exception now or later.

WojciechowskiPiotr replied:
@tbouvet what is your opinion on that? This will affect `docker_swarm` if it starts using `AnsibleDockerSwarmClient` instead of `AnsibleDockerClient`.

I am ok with that proposal, this way we will have one conditional in the constructor and less code in modules checking if we run module on Swarm manager

felixfontein replied:
You could also add a function to `AnsibleDockerSwarmClient` which verifies this, and call that function from this constructor. Then it's easy to add (with almost no code repetition), and `docker_swarm` can only call it if `state` is not `join`.

WojciechowskiPiotr replied:
Moving `check_if_swarm_manager()` to the constructor and implement an exception is fine, all operations except the `join` and `leave` have to be performed on Swarm Manager.

I don't think moving `check_if_swarm_node()` to the constructor is a good approach as what you are checking depends on the operation. It takes as a parameter the node ID, but if empty it is checking the docker host itself. The parameter name differs between modules and for many swarm operations, you don't need to check node. So I'd leave it as it is.

felixfontein replied:
I didn't mean the constructor of `AnsibleDockerSwarmClient`, but the one of a deriving class (resp. in `docker_node_facts`'s main() function, since the `docker_node_fact` module uses `AnsibleDockerSwarmClient` directly).

WojciechowskiPiotr replied:
I am not sure if that's what you recommended:

In `AnsibleDockerSwarmClient` class new method:
```
    def fail_task_if_not_swarm_manager(self):
        if not self.check_if_swarm_manager():
            self.client.fail(msg="This node is not a manager.")
```

In `docker_node_facts` `main()` before getting facts new call:
```
    client.fail_task_if_not_swarm_manager()

    node = get_node_facts(client)
```

In `docker_node`
```
    client.fail_task_if_not_swarm_manager()

    SwarmNodeManager(client, results)
```

Alternatively, we can in the `AnsibleDockerSwarmClient` add this to constructor and make an exceptions for `state` `join` and `leave` as parameters are available for this class. We can even put all logic, not just checking about swarm manager, based on parameters provided to `AnsibleDockerSwarmClient` in constructor.

felixfontein replied:
In `docker_node`, I would call that function from `SwarmNodeManager`'s constructor. Otherwise that looks fine.

I would not have any logic working with `state` option values in `AnsibleDockerSwarmClient`; the client should not depend on how the module's `state` option is designed.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,290 @@
+            Labels=self.parameters.labels,
+        )
+
+        try:
```
Where do you determine that the node's state actually needs to be updated?

WojciechowskiPiotr replied:
I don't think it is necessary because it is up to the internal Docker engine to decide what to do. No matter if you are changing something or not Swarm manager exchanges packets with node. Neither CLI or API informs you that you did not change anything. And I don't see the necessity to add it here, but if we do there should be `force` mode to send the update even if nothing changes.

The `Version`, `Availability` and `Role` are always required, if the `Label` is missing, then labels are removed from the node.

felixfontein replied:
It is customary that Ansible modules return `changed == False` if nothing has to be changed. So you have to determine it anyway. And not doing an API call when you already know that it is not needed is better anyway.

WojciechowskiPiotr replied:
This is not a problem, it has to inspect node before making changes anyway to get up-to-date `Version`

WojciechowskiPiotr replied:
Change tracking added in commit 9e54a068

Commit:
docker_node_facts: update in input and return values, update in documentation section

Commit:
docker_node: Updated operations on labels, tracking if change is required

Commit:
docker_node: Updated documentation, parameter 'hostname' is now required
docker_node_facts: Updated documentation

WojciechowskiPiotr commented:
In the commit 5c691c6 I've changed the `docker_node` parameter `name` to `hostname`. Basically, you can identify Swarm node in two ways - either by `ID` or `Hostname` which is subkey in `Description`. If more than one host is registered with the same `Hostname` then `ID` must be used.

I've changed the parameter name because there is a key `Name` as subkey of `Spec` in node facts and you can update it via API but not via CLI. You also cannot use it in docker-compose when setting up a Swarm service. I don't know the origin and what it is supposed to use for, even the filter `name=something` filters basing on `Hostname` key.

1) I don't think we should implement the manipulation of `Name` key. At least not until there is real usage of it
2) I think using the `name` as a parameter to specify the `hostname` in `docker_node` module might a little misleading in such situation, but same time might be more in line with Ansible best practice of parameters naming. I can change it back or maybe the `id` as a name (accepting both hostname and the node id of course as it is now) would be a better solution?

I will follow on your recommendations on this one.

felixfontein commented:
`hostname` sounds reasonable. We can still add `name` later and make `hostname` non-required.

Commit:
* Failing Ansible tasl if not run on swarm manager - code cleanup

Commit:
* docker_node: Remove the 'action' list from output

WojciechowskiPiotr commented:
ready_for_review

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,301 @@
+        description:
+            - The hostname or ID of node as registered in Swarm.
+            - If more than one node is registered using the same hostname the ID must be used,
+              otherwise task will fail.
```
```suggestion
              otherwise module will fail.
```
You do not control when a task fails. You can only control when this module fails.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,301 @@
+              If no labels are assigned then it will add listed labels. The I(labels) must be specified.
+            - If set as C(merge) for labels that are assigned to node will update their values as specified in I(labels).
+            - If set as C(merge) labels assigned to node that are not specified in I(labels) will remain unchanged.
+            - If set as C(merge) will add new labels specified in I(labels) if labels are not assigned to node.
```
I think it would be better to have one paragraph per possible value.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,301 @@
+            - If set as C(merge) labels assigned to node that are not specified in I(labels) will remain unchanged.
+            - If set as C(merge) will add new labels specified in I(labels) if labels are not assigned to node.
+            - Set to C(replace) to replace assigned labels with provided ones. The I(labels) must be specified.
+            - Set to C(remove) to remove labels specified in I(labels). If I(labels) is not provided then will remove
```
What happens with the values of the labels?

I think it would be better to get rid of `'remove'` and instead add another option, `labels_to_remove`, which is a list (of strings). Also, this option will then be closer to the `comparison` feature of `docker_container`. (Remove all labels can still be done via `labels: {}` and `labels_state: remove`.)

WojciechowskiPiotr replied:
The `comparison` is quite complex and labels manipulation should be much simpler. Extending your proposal how about:

```
- name: Add labels (overwrite the value if label exists)
  docker_node:
    hostname: mynode
    labels_to_add:
      key: value

- name: Remove selected labels
  docker_node:
    hostname: mynode
      labels_to_remove:
      - key1
      - key2

- name: Add and remove labels (if the same label is listed on both options then its value will be updated)
  docker_node:
    hostname: mynode
    labels_to_remove:
      - key1
      - key2
    labels_to_add:
      key: value

- name: Replace labels
  docker_node:
    hostname: mynode
    labels:
      key: value

- name: Remove labels
  docker_node:
    hostname: mynode
    remove_all_labels:
```

So it will give us 4 clear options. Containers are probably the only item where `comparison` fits. This approach is intuitive and gives you the way to manipulate labels even if you don't know the labels assigned to node yet (if you care about them you can just read them first via `docker_node_facts`). 

Or maybe independent module `docker_label` for labels manipulation across all docker modules with those four options?

felixfontein replied:
The comparisons feature for labels is pretty simple: you can replace labels, update labels, or ignore labels.

In genral: Ansible is not about doing explicit operation, but about specifying state and letting Ansible make sure objects adhere to this state. That's why such explicit operation options are not OK in my opinion. Having an option to make sure some specific labels are not there is something I would accept, but not an interface which is different from the general Ansible way.

WojciechowskiPiotr replied:
So to get it all straight we are talking about implementation as below?

```
- name: Add labels (update the value if label exists)
  docker_node:
    hostname: mynode
    labels:
      key: value
    labels_state: merge

- name: Remove selected labels (if exists)
  docker_node:
    hostname: mynode
      labels_to_remove:
      - key1
      - key2

- name: Replace labels
  docker_node:
    hostname: mynode
    labels:
      key: value
    labels_state: replace

- name: Remove labels
  docker_node:
    hostname: mynode
    labels: {}
    labels_state: replace

- name: Add/update and remove at the same time
  docker_node:
    hostname: mynode
    labels:
      key: value
    labels_state: merge
    labels_to_remove:
      - key2
```

In the last example if both non-empty `labels` and `labels_to_remove` are present then if the label is set in both options it will be updated, not removed (it will be clearly stated in the documentation of course) unless the `labels` is `{ }` and `labels_state` is `replace`, in such case it will remove all labels.

felixfontein replied:
Yes, I think this is good. The `labels_to_remove` is kind of an exception; maybe it would also be better to call it `labels_that_should_not_be_there` instead, though that's a pretty horrible name ;) (If you can come up with something better, feel free...)

WojciechowskiPiotr replied:
Might be just `remove_labels` or in less grammatical form `labels_remove`

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,301 @@
+       be installed at the same time. Also note that when both modules are installed
+       and one of them is uninstalled, the other might no longer function and a
+       reinstall of it is required."
+    - "The docker server >= 1.10.0"
```
What do you mean by this?

WojciechowskiPiotr replied:
Fixed, was supposed to be Docker API

felixfontein replied:
I hope the version for Docker API will be different than 1.10.0, too :)

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,301 @@
+        try:
+            status_down = self.client.check_if_swarm_node_is_down()
+        except APIError:
+            return
```
I think you need some proper error handling here.

WojciechowskiPiotr replied:
The `client.check_if_swarm_node_is_down()` calls the `get_node_inspect()` and there is extended handling of API exceptions. I've added here additional `fail()` call here instead of return just to be safe.

felixfontein replied:
If no APIError can leave check_if_swarm_node_is_down, you should remove the try/except :)

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,301 @@
+            return
+
+        if status_down:
+            self.client.fail(msg="Can not update the node. The status node is down.")
```
Shouldn't this either be
```suggestion
            self.client.fail(msg="Can not update the node. The node is down.")
```
or
```suggestion
            self.client.fail(msg="Can not update the node. The node status is down.")
```
?

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,301 @@
+        except APIError as exc:
+            self.client.fail(msg="Failed to get node information for %s" % to_native(exc))
+
+        __changed = False
```
Why not use `changed`? The `__` prefix is only needed for object properties, not for local vaiables.

WojciechowskiPiotr replied:
fixed

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,301 @@
+
+        if __changed is True:
+            try:
+                self.client.update_node(node_id=node_info['ID'], version=node_info['Version']['Index'],
```
You shouldn't call this if check mode is active.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,301 @@
+        )
+
+        if self.parameters.role is None:
+            node_spec['Role'] = node_info['Spec']['Role']
```
If the role changes, you're not setting `__changed` to `True`.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,301 @@
+            node_spec['Role'] = node_info['Spec']['Role']
+
+        if self.parameters.availability is None:
+            node_spec['Availability'] = node_info['Spec']['Availability']
```
Same here.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,301 @@
+
+        if self.parameters.labels_state == 'replace':
+            node_spec['Labels'] = self.parameters.labels
+            __changed = True
```
You need to check whether there is actually a change.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,301 @@
+        if self.parameters.labels_state == 'remove':
+            node_spec['Labels'] = node_info['Spec']['Labels']
+
+            if self.parameters.labels is None and node_info['Spec']['Labels'] is not None:
```
```suggestion
            if self.parameters.labels is None and not node_info['Spec']['Labels']:
```
It could also be empty.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,301 @@
+                node_spec['Labels'] = None
+                __changed = True
+            elif self.parameters.labels is not None and node_info['Spec']['Labels'] is not None:
+                for next_key in self.parameters.labels:
```
Why `next_key` and not simply `key`?

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,301 @@
+                            node_spec['Labels'].pop(next_key)
+                            __changed = True
+                        except KeyError as exc:
+                            self.client.fail(msg="Failed to remove labels for %s" % to_native(exc))
```
Why fail here? If the label wasn't there, there's simply nothing to change.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,301 @@
+                    if self.parameters.labels.get(next_key) == node_info['Spec']['Labels'][next_key]:
+                        pass
+                    else:
+                        node_spec['Labels'].update({next_key: self.parameters.labels.get(next_key)})
```
```suggestion
                        node_spec['Labels'][next_key] = self.parameters.labels.get(next_key)
```

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,301 @@
+                        node_spec['Labels'].update({next_key: self.parameters.labels.get(next_key)})
+                        __changed = True
+                else:
+                    node_spec['Labels'].update({next_key: self.parameters.labels.get(next_key)})
```
```suggestion
                    node_spec['Labels'][next_key] = self.parameters.labels.get(next_key)
```

felixfontein reviewed (commented):
Two general things:
 - please also edit [.github/BOTMETA.yml](https://github.com/ansible/ansible/blob/devel/.github/BOTMETA.yml#L563) and "clone" the docker_common.py entry so that there's also an entry for docker_swarm.py;
 - adding some basic integration tests for both docker_node and docker_node_facts would be nice :)
(See for example the [docker_swarm_service tests](https://github.com/ansible/ansible/tree/devel/test/integration/targets/docker_swarm_service).)

Commit:
* docker_node: variable name change to be align with Python best practice, BOTMETA.yml update

Commit:
* module_utils/docker_swarm.py: fix for incorrect fail() action
* docker_node: documentation and code small updates

Commit:
* docker_node: revised labels manipulation

WojciechowskiPiotr commented:
Pushed the changes that cover all discussed point. Can we move all the integration tests into separate PR (that will also include `docker_swarm_facts` and `docker_host_facts`)?

Commit:
* docker_node: Documentation update

Commit:
* docker_node: Update to node availability and role modification

Commit:
* docker_node: Update to check_mode handling

WojciechowskiPiotr commented:
ready_for_review

Fixed/updated requested parts, changed the approach to labels manipulation, updated BOTMETA, only waiting for a decision on the final name for the option `labels_to_remove`. As mentioned earlier I would like to put all integration tests for all new modules in a separate PR. If there is anything else you think requires an improvement let me know.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -47,9 +51,13 @@
+              names, not their values.
+            - If the label provided on the list is not assigned to the node, the entry is ignored.
+            - If the label is both on the I(labels_to_remove) and I(labels), then value provided in I(labels) remains
+              assigned to the node.
```
Why not fail if a label appears in both `labels` and `labels_to_remove`?

WojciechowskiPiotr replied:
List of labels may be an output of other task or external script. Changing labels assigned to the node usually happens when you do maintenance or emergency operations and in such cases, it is better if one operation (assigning/modifying label) supersede other (removing labels) - module will fail if there is a problem in reading/writing to swarm manager/node (or typo in option). If for someone label removal is more important than preserving it in such situation then the task can be easily split into two separate tasks - first `labels` then `labels_to_remove`
The module may fail here of course, but that means more difficult error handling in the playbook.

felixfontein replied:
Sounds ok. How about still emitting a warning in case a label shows up in both `labels` and `labels_to_remove`? That way users know that something isn't right (and they can investigate), but can also choose to ignore it.

WojciechowskiPiotr replied:
Sure, the warning is a nice solution :)

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,316 @@
+
+    required_if = [
+        ('labels_state', 'merge', ['hostname']),
+        ('labels_state', 'replace', ['hostname']),
```
You can remove the `required_if` entries.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,316 @@
+        type: str
+    labels:
+        description:
+            - User-defined key/value metadata that will be assigned as node attribute. The actual state of labels
```
Maybe start a new paragraph for `The actual state...`.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,316 @@
+    - docker
+requirements:
+    - "python >= 2.6"
+    - "docker-py >= 1.10.0"
```
Some of the features probably require a newer docker-py version and/or API version. Let's take a look at that later, though, when a first version is merged (see #51175 for example).

WojciechowskiPiotr replied:
Sure

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,316 @@
+    labels:
+      key: value
+
+- name: Merge node labels and new labels
```
I don't think you need two examples for the same thing. That's just making it harder to find examples. Why not merge them together and add a comment before `labels_state` like `# The following isn't necessary, since merge is the default value:`?

WojciechowskiPiotr replied:
Or maybe just leave the example without the `labels_state: merge`? The default value is described in the documentation clearly

felixfontein replied:
Yeah, or that.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,316 @@
+  docker_node:
+    hostname: mynode
+    labels:
+      key: value
```
How about:
```suggestion
      # If key was defined before, its value will be updated.
      # If it was not defined before, this label will be added.
      key: value
```

WojciechowskiPiotr replied:
Should we duplicate the information that already is provided in `labels_state` description?

WojciechowskiPiotr replied:
I think the operation on labels is clearly described in the description of `labels_state` so let's not add comments here.

felixfontein replied:
Fine for me. Sometimes it's nicer to make examples more explicit, but that was just a suggestion, so I don't mind leaving it away :)

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node_facts.py:
```diff
@@ -29,8 +29,20 @@
       - The name of the node to inspect.
-      - When identifying an existing node name may either the hostname of the node (as registered in Swarm) or node ID.
-    required: true
+      - The list of nodes names to inspect.
```
I think it would be better to do the functionality change for `docker_node_facts` in a separate PR, so that this PR only adds the module_utils module and `docker_node` (and changes `docker_node_facts` so it uses the new module_utils module).

WojciechowskiPiotr replied:
I already have `docker_swarm_facts` PR in the queue that adopts the module to use the new module_utils module so I think it will be better move all the changes to `docker_node_facts` to separate PR to keep consistent.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,316 @@
+    ]
+
+    option_minimal_versions = dict(
+        signing_ca_cert=dict(docker_api_version='1.30'),
```
You can remove these entries, since these are not options of the `docker_node` module. Keep the `dict`, though, since we'll probably need it for labels (and potentially some more options).

WojciechowskiPiotr replied:
I removed `option_minimal_versions` and changed `min_docker_api_version` from `1.24` to `1.25` because this version allows using both `id` and `name` to identify node when updating the node, the `1.24` accepts only `id` for this operation.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,316 @@
+                node_spec['Labels'] = self.parameters.labels
+                changed = True
+
+        if self.parameters.labels_state == 'merge':
```
To make it clearer this belongs to the `if` above:
```suggestion
        elif self.parameters.labels_state == 'merge':
```
Also maybe remove the empty line between them.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,316 @@
+                    changed = True
+            else:
+                node_spec['Labels'] = self.parameters.labels
+                changed = True
```
You need to check whether something actually changes through this before setting `changed` to `true`.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,316 @@
+        if self.parameters.labels_state == 'replace':
+            if self.parameters.labels is None:
+                node_spec['Labels'] = {}
+                if node_info['Spec']['Labels'] is not None:
```
It could not be `None`, but an empty `dict`. So it's better to write
```suggestion
                if node_info['Spec']['Labels']:
```

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,316 @@
+                changed = True
+
+        if self.parameters.labels_state == 'merge':
+            node_spec['Labels'] = node_info['Spec']['Labels']
```
You should do
```suggestion
            node_spec['Labels'] = dict(node_info['Spec']['Labels'] or {})
```
to a) make sure that you work with a copy, and b) to make sure that it is not `None`.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,316 @@
+            node_spec['Labels'] = node_info['Spec']['Labels']
+            if self.parameters.labels is not None:
+                for key in self.parameters.labels:
+                    if key in node_info['Spec']['Labels']:
```
Make sure this is not `None`.

felixfontein replied:
Instead of this whole block (`if / if / else / else`), you could also simply compare `node_info['Spec']['Labels'].get(key)` to `value`. That's one simple `if`, then.

WojciechowskiPiotr replied:
I've rewritten the whole section as per your three recommendations to simplify it and `.get()` in `if` handles exceptions.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,316 @@
+        if self.parameters.labels_state == 'merge':
+            node_spec['Labels'] = node_info['Spec']['Labels']
+            if self.parameters.labels is not None:
+                for key in self.parameters.labels:
```
```suggestion
                for key, value in self.parameters.labels.items():
```
Then you can write `value` instead of `self.parameters.labels.get(key)` below.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,316 @@
+            if self.parameters.labels is not None:
+                for key in self.parameters.labels:
+                    if key in node_info['Spec']['Labels']:
+                        if self.parameters.labels.get(key) == node_info['Spec']['Labels'][key]:
```
Why not
```suggestion
                        if self.parameters.labels.get(key) != node_info['Spec']['Labels'][key]:
```
and get rid of the `pass / else`?

Commit:
* docker_node: Code cleanup

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,294 @@
+                if node_info['Spec']['Labels']:
+                    changed = True
+            else:
+                if node_info['Spec']['Labels'].items() != self.parameters.labels.items():
```
There's no need to use `items()`, in fact it might give wrong results (since now you're comparing two iterables or list of items, depending on Python version, instead of comparing two dictionaries). Also, this won't work if `node_info['Spec']['Labels']` is `None`.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,294 @@
+            node_spec['Labels'] = dict(node_info['Spec']['Labels'] or {})
+            if self.parameters.labels is not None:
+                for key, value in self.parameters.labels.items():
+                    if node_info['Spec']['Labels'].get(key) != value:
```
How about
```suggestion
                    if node_spec['Labels'].get(key) != value:
```
That won't die if `node_info['Spec']['Labels']` is `None`.

WojciechowskiPiotr replied:
Right, line 236 handles that

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,294 @@
+                            changed = True
+                    else:
+                        self.client.module.warn(
+                            "Label %s listed both in 'labels' and 'labels_to_remove'. Keeping the assigned label value."
```
I would quote `%s`, since labels can be arbitrary strings.

WojciechowskiPiotr replied:
The only allowed (but not enforced) are alphanumeric, dot and hypen. You recommend this change to improve the warning readability?:
```
[WARNING]: Label 'key-with-hypens' listed both in 'labels' and 'labels_to_remove'. Keeping the assigned label value.
```

I think the `to_native()` takes care of proper for different Python version variable output, but I can always try to use `str()`.

felixfontein replied:
Yeah, it's mostly readability. With the quotes, it's easier to parse. Especially if the label is a word which could be parsed as part of the sentence. Just think of when someone uses the label `was` :-)

Also, `to_native()` should be fine, or just writing `... % key`.

WojciechowskiPiotr replied:
So it is already covered in latest commits :)

Commit:
docker_node_facts: Code cleanup

felixfontein commented on .github/BOTMETA.yml:
```diff
@@ -570,6 +570,9 @@ files:
     maintainers: $team_docker
     supershipit: felixfontein
+  $module_utils/docker_swarm.py:
+    maintainers: $team_docker
```
In case you want to be informed as well for PRs and issues involving module_utils/docker_swarm.py, you can also add yourself:
```suggestion
    maintainers: $team_docker WojciechowskiPiotr
```
(Alternatively, you can add yourself to the $team_docker definition at the bottom of the file if you are interested on other docker_* modules as well. If you don't do that, you're only informed when one of the modules you are listed as author for is affected.)

WojciechowskiPiotr replied:
I'll add myself to $team_docker as I don't mind participating actively

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,295 @@
+       reinstall of it is required."
+    - Docker API >= 1.25
+author:
+  - Piotr Wojciechowski (@wojciechowskipiotr)
```
```suggestion
  - Piotr Wojciechowski (@WojciechowskiPiotr)
```
ansibot can get confused if the case doesn't match.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,295 @@
+                if node_info['Spec']['Labels']:
+                    changed = True
+            else:
+                if node_info['Spec']['Labels'] != self.parameters.labels:
```
```suggestion
                if (node_info['Spec']['Labels'] or {}) != self.parameters.labels:
```
Just in case `node_info['Spec']['Labels']` happens to be `None`.

felixfontein reviewed (commented):
Looks good. Can you re-add the change to `docker_node_facts` which only affect using the new module_utils module, and which does not change any functionality?

Commit:
docker_node_facts: Adding back the module with only update to use AnsibleDockerSwarmClient instead of AnsibleDockerClient
docker_node: cosmetic code changes
BOTMETA: updated on $team_docker

WojciechowskiPiotr commented:
> Looks good. Can you re-add the change to `docker_node_facts` which only affect using the new module_utils module, and which does not change any functionality?

Done

WojciechowskiPiotr commented:
As there was no additional feedback for the past three days can we move forward with this and send it for core team review?

WojciechowskiPiotr commented:
shipit

felixfontein commented:
@WojciechowskiPiotr due to a change to BOTMETA (#51544) this needs to be rebased. Unfortunately that other PR exposed a bug in ansibot (ansible/ansibullbot#1155), so it could happen that BOTMETA will be changed again soonish (which might require another rebase). Anyway, we can try to get this merged before this happens, but first you need to rebase :)

Commit:
docker_node: BOTMETA update

gundalow commented on .github/BOTMETA.yml:
```diff
@@ -584,6 +584,7 @@ files:
     maintainers: $team_windows_core
     support: core
   $module_utils/docker_common.py: *docker
+  $module_utils/docker_swarm.py: *docker
```
Thanks for remembering to update this, if you replace the above line with:
```
  $module_utils/docker: *docker
```
We can future proof this.

gundalow commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,295 @@
+extends_documentation_fragment:
+    - docker
+requirements:
+    - "python >= 2.6"
```
Ansible 2.8 only supports Python 2.6+ on the remote, so this line isn't needed.

gundalow commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,295 @@
+    - "Please note that the L(docker-py,https://pypi.org/project/docker-py/) Python
+       module has been superseded by L(docker,https://pypi.org/project/docker/)
+       (see L(here,https://github.com/docker/docker-py/issues/1310) for details).
+       For Python 2.6, C(docker-py) must be used. Otherwise, it is recommended to
```
Ansible 2.8 only supports Python 2.6+ on the remote, so this could be reworded.

felixfontein replied:
Let's keep this for now, and let's change all of these in a subsequent PR for all `docker_*` modules, so that all modules get the same text.

gundalow replied:
Agreed, follow up PR is a good idea.

gundalow commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,295 @@
+'''
+
+RETURN = '''
+node_facts:
```
If we know the structure of the `dict` then adding some of those details would be good, see the example at https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#return-block

felixfontein replied:
We only know a subset of the structure, as newer API versions will add more data here. For other docker_* modules, we "solved" this by adding `sample:` with an excerpt of a result containing the most interesting things. No idea what's the best approach here.

(Another thing is that depending on the data returned, this can be *really* complex. For docker_container, the returned inspect result is a huge dict; documenting everything would be a big task.)

WojciechowskiPiotr replied:
Maybe a direct link to the latest Docker API documentation instead if `sample:` and information to check the API documentation for the specific API version on remote?

felixfontein replied:
That would be fine, too. Maybe also have `sample:` as well, so people get an idea what they will get.

gundalow commented:
@WojciechowskiPiotr Thank you for this PR
@felixfontein Thank you for the review

felixfontein commented:
@WojciechowskiPiotr thanks for the PR! You will notice that you'll get CC'ed in all docker issues/PRs; ansibot will take some time to add you to all existing issues/PRs.
@tbouvet thanks for your feedback!
@gundalow thanks for reviewing and merging!

I've opened a first follow-up PR in #51700 with some things discussed with @gundalow.

felixfontein commented:
@WojciechowskiPiotr thanks for the PR! You will notice that you'll get CC'ed in all docker issues/PRs; ansibot will take some time to add you to all existing issues/PRs.
@tbouvet thanks for your feedback!
@gundalow thanks for reviewing and merging!

I've opened a first follow-up PR in #51700 with some things discussed with @gundalow.

felixfontein commented:
More stuff for follow-up PRs (before I forget about this):
- integration tests (for both docker_node and docker_node_facts)
- minimal version checking for docker-py and docker API versions
- other changes to docker_node_facts which were temporarily in this PR

@WojciechowskiPiotr anything else? :)

WojciechowskiPiotr commented:
@felixfontein docker_swarm_facts from #50622

felixfontein commented:
@WojciechowskiPiotr that already has a PR (and is in the tab to the left of this PR in my Ansible docker tab collection ;-) ), so I didn't mention it.
